controllers/machine_controller.go

```diff
@@ -534,6 +544,31 @@ func (r *MachineReconciler) drainNode(ctx context.Context, cluster *clusterv1.Cl
 	return ctrl.Result{}, nil
 }
 
+// pod deletion and volume detach happen asynchronously, so pod could be deleted before volume detached from the node
+// this could cause issue for some storage provisioner, for example, vsphere-volume this is problematic
+// because if the node deleted before detach success, then the underline VMDK will be deleted together with the Machine
+// so after node draining we need to check if all volumes are detached before deleting the node.
+func (r *MachineReconciler) isNodeVolumesDetached(ctx context.Context, cluster *clusterv1.Cluster, nodeName string, machineName string) (bool, error) {
+	log := ctrl.LoggerFrom(ctx, "cluster", cluster.Name, "node", nodeName, "machine", machineName)
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		log.Error(err, "Error creating a remote client for cluster while waiting for volume detach")
+		return false, err
+	}
+
+	node := &corev1.Node{}
+	if err = remoteClient.Get(ctx, types.NamespacedName{Name: nodeName}, node); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Error(err, "Could not find node from noderef, it may have already been deleted")
+			return true, nil
```

Comment:
The idea is to ignore the error so that the machine deletion isn't blocked because of this? 

I would have assumed that this case should never happen as the node should always still exist at this time.

Reply:
Agree that it should not happen.

Reply:
I wonder if we should really ignore the error in this case. But I have no strong opinion on that, maybe somebody else has.

Reply:
We should ignore and continue with execution if the node was deleted out of band